src/sagemaker/serve/validations/optimization.py

```diff
@@ -0,0 +1,206 @@
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+"""Holds the validation logic used for the .optimize() function. INTERNAL only"""
+from __future__ import absolute_import
+
+import textwrap
+import logging
+from typing import Any, Dict, Set, Optional
+from enum import Enum
+from pydantic import BaseModel
+
+logger = logging.getLogger(__name__)
+
+
+class _OptimizationContainer(Enum):
+    """Optimization containers"""
+
+    TRT = "TRT"
+    VLLM = "vLLM"
+    NEURON = "Neuron"
+
+
+class _OptimizationCombination(BaseModel):
+    """Optimization ruleset data structure for comparing input to ruleset"""
+
+    optimization_container: _OptimizationContainer = None
+    compilation: Set[Optional[bool]]
+    speculative_decoding: Set[Optional[bool]]
+    sharding: Set[Optional[bool]]
+    quantization_technique: Set[Optional[str]]
+
+    def validate_against(self, optimization_combination, rule_set: _OptimizationContainer):
+        """Validator for optimization containers"""
+
+        # check the validity of each individual field
+        if not optimization_combination.compilation.issubset(self.compilation):
+            raise ValueError("Compilation")
+        if not optimization_combination.quantization_technique.issubset(
+            self.quantization_technique
+        ):
+            copy_quantization_technique = optimization_combination.quantization_technique.copy()
+            raise ValueError(f"Quantization:{copy_quantization_technique.pop()}")
+        if not optimization_combination.speculative_decoding.issubset(self.speculative_decoding):
+            raise ValueError("Speculative Decoding")
+        if not optimization_combination.sharding.issubset(self.sharding):
+            raise ValueError("Sharding")
+
+        # optimization technique combinations that need to be validated
+        if optimization_combination.compilation and optimization_combination.speculative_decoding:
+            is_compiled = optimization_combination.compilation.copy().pop()
+            is_speculative_decoding = optimization_combination.speculative_decoding.copy().pop()
+            if is_compiled and is_speculative_decoding:
+                raise ValueError("Compilation and Speculative Decoding together")
+
+        if rule_set == _OptimizationContainer.TRT:
+            is_compiled = optimization_combination.compilation.copy().pop()
+            is_quantized = optimization_combination.quantization_technique.copy().pop()
+            if is_compiled and not is_quantized or is_quantized and not is_compiled:
+                raise ValueError(f"Compilation must be provided with Quantization")
```

Comment:
Compilation stand alone should be supported.

Reply:
Done